### PR TITLE
Add noisy channel: thermal relaxation

### PR DESCRIPTION
**Pull request by alejomonbar**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Adding a new noisy channel which gives the thermal relaxation channel as part of the #971 issue. I got the information from the [supplementary](https://static-content.springer.com/esm/art%3A10.1038%2Fs41534-020-0272-6/MediaObjects/41534_2020_272_MOESM1_ESM.pdf) of [Quantum classifier with tailored quantum kernel](https://www.nature.com/articles/s41534-020-0272-6.pdf). I still need to figure out how to get the second case when T2 > T1.
**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by josh146** [comment]
Thanks @alejomonbar for this nice contribution! 🙂 Let us know if you have any questions, or when this PR is ready for review, and someone from the team will be on hand to help.

**Comment by alejomonbar** [comment]
> @alejomonbar I checked the different documents about thermal relaxation. As PennyLane has no Choi representation of quantum channel, we indeed need to construct the kraus matrices from the Choi matrix given in the paper. Therefore for the case where dephasing is more prevalent than relaxation (T2>T1), you will need indeed need this transformation. For the moment you can create this function directly in `channel.py` and work there. I will check if a better location is needed.
> 
> Thank you again for your contribution and feel free to ask any question. Let me know when the PR will be ready for review!

Thank you, I will look at that and come back when I implement it!

**Comment by alejomonbar** [comment]
> Thank you for the contribution, before I can give a full review you have different points to work on:
> 
> 1. Codefactor: in the folder `channel.py` minor things like trailinng white spaces and too many statements should be solved easily.
> 2. You can `pip install black` and run `black -l 100 your_file.py` on all the changed files.
> 3. For docstrings and documentation, there probably some small errors of formatting.
> 4. Some tests are failing, like `test_kraus_matrices_sum_identity`did you check it locally? Maybe there is small mistake in the Kraus operators or you need to adapt the test for the case. Same for the new tests are failing as well, do they pass locally for you?

Thank you @rmoyard , I tried it but I have a problem when trying to run the test I got the following message, `ModuleNotFoundError:No module named 'autoray'` and I installed it using `pip install autoray`, and looks like it is installed, but still, I got the same message.

**Comment by rmoyard** [comment]
> > Thank you for the contribution, before I can give a full review you have different points to work on:
> > 
> > 1. Codefactor: in the folder `channel.py` minor things like trailinng white spaces and too many statements should be solved easily.
> > 2. You can `pip install black` and run `black -l 100 your_file.py` on all the changed files.
> > 3. For docstrings and documentation, there probably some small errors of formatting.
> > 4. Some tests are failing, like `test_kraus_matrices_sum_identity`did you check it locally? Maybe there is small mistake in the Kraus operators or you need to adapt the test for the case. Same for the new tests are failing as well, do they pass locally for you?
> 
> Thank you @rmoyard , I tried it but I have a problem when trying to run the test I got the following message, `ModuleNotFoundError:No module named 'autoray'` and I installed it using `pip install autoray`, and looks like it is installed, but still, I got the same message.

Does it appear in the environment if you write `pip list`?

**Comment by alejomonbar** [comment]
> > > Thank you for the contribution, before I can give a full review you have different points to work on:
> > > 
> > > 1. Codefactor: in the folder `channel.py` minor things like trailinng white spaces and too many statements should be solved easily.
> > > 2. You can `pip install black` and run `black -l 100 your_file.py` on all the changed files.
> > > 3. For docstrings and documentation, there probably some small errors of formatting.
> > > 4. Some tests are failing, like `test_kraus_matrices_sum_identity`did you check it locally? Maybe there is small mistake in the Kraus operators or you need to adapt the test for the case. Same for the new tests are failing as well, do they pass locally for you?
> > 
> > 
> > Thank you @rmoyard , I tried it but I have a problem when trying to run the test I got the following message, `ModuleNotFoundError:No module named 'autoray'` and I installed it using `pip install autoray`, and looks like it is installed, but still, I got the same message.
> 
> Does it appear in the environment if you write `pip list`?

Yes, it does with version `autoray 0.2.5`

**Comment by rmoyard** [comment]
@alejomonbar Could you give me your system information? `import pennylane as qml; qml.about()`

File: pennylane/ops/channel.py [hunk 1]
```diff
@@ -466,6 +466,148 @@ def _kraus_matrices(cls, *params):
...
+                1 & 0
+                \end{bmatrix}
+
+    .. math::
+        K_5 = \sqrt{p_{r1}}\begin{bmatrix}
+                0 & 0  \\
+                0 & 1
+                \end{bmatrix}
+
+    where :math:`p_{r0} \in [0, 1]` is the probability of a reset to 0, :math:`p_{r1} \in [0, 1]` is the probability of
```
Comment (rmoyard) [inline pennylane/ops/channel.py:513]:
```suggestion
    where :math:`p_{r0} \in [0, 1]` is the probability of a reset to 0, :math:`p_{r1} \in [0, 1]` is the probability of
```

File: pennylane/ops/channel.py [hunk 2]
```diff
@@ -466,6 +466,148 @@ def _kraus_matrices(cls, *params):
...
+        K_5 = \sqrt{p_{r1}}\begin{bmatrix}
+                0 & 0  \\
+                0 & 1
+                \end{bmatrix}
+
+    where :math:`p_{r0} \in [0, 1]` is the probability of a reset to 0, :math:`p_{r1} \in [0, 1]` is the probability of 
+    a reset to 1 error, :math:`p_z \in [0, 1]` is the probability of a phase flip (Pauli :math:`Z`) error.
+
+    Case :math:`T_2 > T_1`:
+
```
Comment (rmoyard) [inline pennylane/ops/channel.py:517]:
```suggestion

```

File: pennylane/ops/channel.py [hunk 3]
```diff
@@ -466,6 +466,148 @@ def _kraus_matrices(cls, *params):
...
+                        1 - p_e * p_{reset} & 0 & 0 & eT_2 \\
+                        0 & p_e * p_{reset} & 0 & 0 \\
+                        0 & 0 & (1 - p_e) * p_{reset} & 0 \\
+                        eT_2 & 0 & 0 & 1 - (1 - p_e) * p_reset
+                        \end{bmatrix}
+
+    .. math::
+        K_N = \sqrt{\lambda} \Phi(\nu_{\lambda})
+
+    where :math:`\lambda` are the eigenvalues of the Choi matrix, :math:`\nu_{\lambda}` are the eigenvectors of
```
Comment (rmoyard) [inline pennylane/ops/channel.py:531]:
```suggestion
    where :math:`\lambda` are the eigenvalues of the Choi matrix, :math:`\nu_{\lambda}` are the eigenvectors of
```

File: pennylane/ops/channel.py [hunk 4]
```diff
@@ -466,6 +466,148 @@ def _kraus_matrices(cls, *params):
...
+                        0 & p_e * p_{reset} & 0 & 0 \\
+                        0 & 0 & (1 - p_e) * p_{reset} & 0 \\
+                        eT_2 & 0 & 0 & 1 - (1 - p_e) * p_reset
+                        \end{bmatrix}
+
+    .. math::
+        K_N = \sqrt{\lambda} \Phi(\nu_{\lambda})
+
+    where :math:`\lambda` are the eigenvalues of the Choi matrix, :math:`\nu_{\lambda}` are the eigenvectors of 
+    the choi_matrix, and :math:`\Phi(x)` is a isomorphism from :math:`\mathbb{C}^{n^2}`
```
Comment (rmoyard) [inline pennylane/ops/channel.py:532]:
```suggestion
    the choi_matrix, and :math:`\Phi(x)` is a isomorphism from :math:`\mathbb{C}^{n^2}`
```

File: pennylane/ops/channel.py [hunk 5]
```diff
@@ -466,6 +466,148 @@ def _kraus_matrices(cls, *params):
...
+    Case :math:`T_2 > T_1`:
+    
+    The Choi matrix is given by
+
+    .. math::
+        \Lambda = \begin{bmatrix}
+                        1 - p_e * p_{reset} & 0 & 0 & eT_2 \\
+                        0 & p_e * p_{reset} & 0 & 0 \\
+                        0 & 0 & (1 - p_e) * p_{reset} & 0 \\
+                        eT_2 & 0 & 0 & 1 - (1 - p_e) * p_reset
```
Comment (rmoyard) [inline pennylane/ops/channel.py:525]:
```suggestion
                        eT_2 & 0 & 0 & 1 - (1 - p_e) * p_{reset}
```

**Review by rmoyard — COMMENTED** [review]
It looks good to me, last comments for Code factor errors and also a dosctring. After that it should be all good 💯

**Pull request merged**